src/toil/spark/spark_cluster.py

```diff
@@ -0,0 +1,439 @@
+                                             check_output = True)[:-1]
+        _log.info("Started HDFS Datanode.")
+        self.hdfsContainerID = _docker_call(no_rm = True,
+                                            work_dir = os.getcwd(),
+                                            tool = "quay.io/ucsc_cgl/apache-hadoop-master:2.6.2",
+                                            docker_parameters = ["--net=host",
+                                                                 "-d"],
+                                            tool_parameters = [self.hostname],
+                                            sudo = self.sudo,
+                                            check_output = True)[:-1]
+
+        return self.hostname
+
+
+    def stop(self, fileStore):
+        """
+        Stop and remove spark and hdfs master containers
+
+        fileStore: Unused
+        """
+
+        print >> _fp, "Stopping master"
+        
+        sudo = []
+        if self.sudo:
+            sudo = ["sudo"]
+            
+        subprocess.call(sudo + ["docker", "exec", self.sparkContainerID, "rm", "-r", "/ephemeral/spark"])
+        subprocess.call(sudo + ["docker", "stop", self.sparkContainerID])
+        subprocess.call(sudo + ["docker", "rm", self.sparkContainerID])
+        _log.info("Stopped Spark master.")
+
+        subprocess.call(sudo + ["docker", "stop", self.hdfsContainerID])
+        subprocess.call(sudo + ["docker", "rm", self.hdfsContainerID])
+        _log.info("Stopped HDFS namenode.")
+
+        print >> _fp, "Stopped master"
+
+        return
+
+
+    def check(self):
+        """
+        Checks to see if Spark master and HDFS namenode are still running.
+        """
+        
+        status = _checkContainerStatus(self.sparkContainerID, self.hdfsContainerID)
+        print >> _fp, "Master status: %s" % status
+
+        return status
+
+class WorkerService(Job.Service):
+    """
+    Service Job that implements the worker nodes in a Spark/HDFS cluster.
+    Should not be called outside of `SparkService`.
+    """
+    
+    def __init__(self, masterIP, sudo, memory, cores = None):
+        """
+        :param sudo: Whether this code should run docker containers with sudo.
+        :param memory: The amount of memory in Gibibytes that should be \
+        allocated for each node in the spark cluster.
+        :param cores: Optional parameter to set the number of cores per node. \
+        If not provided, we use the number of cores on the node that launches \
+        the service.
+        :type sudo: boolean
+        :type memory: int
+        :type cores: boolean
+        """
+
+        self.masterIP = masterIP
+        self.sudo = sudo
+        self.memory = memory
+
+        if cores:
+            self.cores = cores
+        else:
+            self.cores = multiprocessing.cpu_count()
+            
+        print >> _fp, "Adding worker service with %s memory %s cores" % (self.memory,
```

Comment:
Please use the logger for this and other standard error reporting. Despite annoyances with the unit test framework - we should fix that rather than do this.